### PR TITLE
Add Spi support

### PR DESCRIPTION
**Pull request by folkertdev**
This PR adds SPI communication support.

This work was done at @tweedegolf while working on a commercial project, and we'd now like open source it. It currently runs in production.

Clearly, adding a whole new communication protocol is a pretty invasive change. Here I'm showing the version that we ended up using in our projects, but that does not mean it's the best approach for `lis3dh-rs`  in general. We're open to modifying this code to make it fit.

At a high level

- common functionality is put in the `Lis3dhImpl` trait. For all but 3 functions there are default implementations. This allows us to share a lot of code between the I2C and SPI implementations
- i2c and spi have their own modules, with their own `Lis3dh` struct and implementations of the `Lis3dhImpl` trait and accelerometer traits. 
- The `Error` type needs 2 type paramters now, one for bus and one for pin errors.

Some open questions

- the `Lis3dhImpl` trait leaks more implementation details than the current version. Maybe that's fine, maybe we should not expose that trait but make a supertrait that only exposes the desired methods.
- testing with `embedded_hal_mock`
- there may be some tricks that can give us just one `Lis3dh` struct with `new_i2c()` and `new_spi` functions. Not sure if that's actually nicer in practice though

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -372,37 +383,37 @@ where
...
-
-        let x = i16::from_le_bytes(accel_bytes[0..2].try_into().unwrap());
-        let y = i16::from_le_bytes(accel_bytes[2..4].try_into().unwrap());
-        let z = i16::from_le_bytes(accel_bytes[4..6].try_into().unwrap());
+pub struct Configuration {
+    /// When set, overrides block_data_update
+    pub enable_temperature: bool,
+    /// Whether data should only be updated after reading
+    pub block_data_update: bool,
+    pub mode: Mode,
```
Comment (sajattack) [inline src/lib.rs:685]: Please add doc comments for the rest of these.

**Comment by sajattack** [comment]
Looks great, thanks for contributing back :smile: Just needs a bit more documentation I think.

**Comment by sajattack** [comment]
@jacobrosenthal @BenBergman would either of you like to review this as well?

File: src/i2c.rs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    ) -> Result<u8, Error<Self::BusError, Self::PinError>> {
+        let mut data = [0];
+
+        self.i2c
+            .write_read(self.address, &[register.addr()], &mut data)
+            .map_err(Error::Bus)
+            .and(Ok(data[0]))
+    }
+}
+
```
Comment (sajattack) [inline src/i2c.rs]: Are these exposed now with your changes?

Comment (sajattack) [reply src/i2c.rs]: Maybe make a trait called Lis3dhRaw or somesuch to hide these.

**Comment by folkertdev** [comment]
yes those are now exposed. That's why I kept playing around a bit, and I found this alternative solution that uses just one `Lis3dh` struct, and you can pick either spi or i2c based on using `new_i2c` or `new_spi`. I personally like this design better exactly because it gives better control about what is exposed, and because everything can be nicely exposed from just one module. 

[the alternative implementation](https://github.com/BenBergman/lis3dh-rs/compare/master...folkertdev:one-struct-spi)

In this case, there is the `Lis3dhCore` trait (though `Raw` might be a better name) that gets implemented by a struct wrapping an I2C device, and by a struct wrapping an SPI device. Then the main `Lis3dh` struct demands an argument that implements this `Core` trait. in its constructors. 

Downside is that the error type is now always an associated type, so it's harder to figure out what exactly is in an error.

What are your thoughts on the tradeoffs here?

**Comment by sajattack** [comment]
Sorry, I don't understand what you mean by this
> Downside is that the error type is now always an associated type, so it's harder to figure out what exactly is in an error.

**Comment by folkertdev** [comment]
actually, nevermind. What I thought was a problem is really just more of what this library already has. It's generic over the error type that the I2C device may have. My worry was that it would be hard to figure out what type the error has (and usually, at least on the devices that I work on, a `Debug` implementation will not save you. I generally need `defmt::Format`). 

In this PR, instead we have errors of the form `Error<Self::BusError, Self::PinError>`. What type `BusError` has is not immediately clear, especially when the `Lis3dhCore` trait is not publicly exposed. But I guess this is not too different from not really knowing the concrete I2C error type.

**Comment by folkertdev** [comment]
OK, I moved this over to the new, better, one-file approach, added more doc comments, and even some full configuration examples. Let me know what you think.

We also have interrupt configuration working, which I can add as a separate PR.

**Comment by sajattack** [comment]
Remind me if I haven't touched this by Monday, I want to take some time to try it out.

**Comment by sajattack** [comment]
Seems like i2c still works anyway, don't have spi to test with :+1:

**Comment by sajattack** [comment]
Published as crate version 0.4.1. I accidentally published 0.4.0 with no changes since 0.3.0 and I don't think I have a way to fix that. I pulled the wrong git remote.

**Pull request merged**